### PR TITLE
[renault] Refactor/improve logging and error handleing. New channels …

### PR DESCRIPTION
**Pull request by dougculnane**
As a result of discussion in the forum starting here:
https://community.openhab.org/t/betatest-renault-ze-services-binding/72226/212

The following has been improved in the Renault binding.

- refactor the error logging for clarity and consistency.
- disable features that do not work on certain cars and reduce repeated logging of errors.
- add 3 new channels.
- improve documentation of channels.
- update sitemap

## New Channels

| Channel ID             | Type               | Description                                     | Read Only |
|------------------------|--------------------|-------------------------------------------------|-----------|
| batterystatusupdated   | DateTime           | Timestamp of the last battery status update     | Yes       |
| lockstatus             | String             | Lock status of the car                          | Yes       |
| updatenow              | Switch             | Switch to force an immediate update of channels | No        |

I and user NikM have tested this and it seams stable.  The lock status does not work on our cars but it should work if this documentation is correct. https://renault-api.readthedocs.io/en/latest/endpoints.html#lock-status

### CODE REVIEW

File: bundles/org.openhab.binding.renault/README.md [hunk 1]
```diff
@@ -19,22 +19,23 @@ No discovery
...
+| Parameter         | Description                                                                | Required                |
+|-------------------|----------------------------------------------------------------------------|-------------------------|
+| myRenaultUsername | MyRenault Username.                                                        | Yes                     |
+| myRenaultPassword | MyRenault Password.                                                        | Yes                     |
+| locale            | MyRenault Location (language_country).                                     | Yes                     |
+| vin               | Vehicle Identification Number.                                             | Yes                     |
+| refreshInterval   | Interval the car is polled in minutes.                                     | Yes (default available) |
+| updateDelay       | How long to wait for commands to reach car and update to server in seconds.| Yes (default available) |
+| kamereonApiKey    | Kamereon API Key.                                                          | Yes (default available) |
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/README.md:30]: I think "no" here was fine. Perhaps you could add the default values to the description instead?

File: bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/api/MyRenaultHttpSession.java [hunk 2]
```diff
@@ -233,108 +241,101 @@ public void getLocation(Car car)
...
+    }
+
+    private void checkResponse(ContentResponse response)
+            throws RenaultForbiddenException, RenaultNotImplementedException {
+        switch (response.getStatus()) {
+            case HttpStatus.FORBIDDEN_403:
+                throw new RenaultForbiddenException(
+                        "Kamereon request forbidden! Ensure the car is paired in your MyRenault App.");
+            case HttpStatus.NOT_FOUND_404:
+                throw new RenaultNotImplementedException("Kamereon serivce not found");
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/api/MyRenaultHttpSession.java:329]:
```suggestion
                throw new RenaultNotImplementedException("Kamereon service not found");
```

File: bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/handler/RenaultHandler.java [hunk 3]
```diff
@@ -196,6 +200,14 @@ public void handleCommand(ChannelUID channelUID, Command command) {
                         return;
                     }
                 }
+            case RenaultBindingConstants.CHANNEL_UPDATE_NOW:
+                if (command instanceof OnOffType && OnOffType.valueOf(command.toString()).equals(OnOffType.ON)) {
+                    if (pollingJob != null) {
+                        pollingJob.cancel(true);
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/handler/RenaultHandler.java:206]: Looks like this will introduce a null annotation warning?

File: bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/handler/RenaultHandler.java [hunk 4]
```diff
@@ -196,6 +200,14 @@ public void handleCommand(ChannelUID channelUID, Command command) {
                         return;
                     }
                 }
+            case RenaultBindingConstants.CHANNEL_UPDATE_NOW:
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/handler/RenaultHandler.java:203]: Why not implement `RefreshType` instead?

File: bundles/org.openhab.binding.renault/src/main/resources/OH-INF/i18n/renault_fr.properties [hunk 5]
```diff
@@ -55,6 +55,9 @@ thing-type.config.renault.car.vin.description = Numéro d'identification du véh
 # channel types
 
 channel-type.renault.batteryavailableenergy.label = Énergie Disponible Batterie
+channel-type.renault.batterystatusupdated.label = État de la batterie mis à jour
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/src/main/resources/OH-INF/i18n/renault_fr.properties:58]: Please remove translation changes from the PR and provide them in Crowdin after the PR has been merged. See https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations

File: bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml [hunk 6]
```diff
@@ -94,7 +97,13 @@
 	<channel-type id="batteryavailableenergy">
 		<item-type>Number:Energy</item-type>
 		<label>Battery Energy Available</label>
-		<state pattern="%.1f %unit%" readOnly="true"></state>
+		<state pattern="%.1f kWh" readOnly="true"></state>
+	</channel-type>
+	<channel-type id="batterystatusupdated">
+		<item-type>DateTime</item-type>
+		<label>Battery Status Updated</label>
+		<description>Timestamp of the last battery status update</description>
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml:105]: Category could also be added:
```suggestion
		<description>Timestamp of the last battery status update</description>
		<category>Time</category>
```

File: bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml [hunk 7]
```diff
@@ -94,7 +97,13 @@
 	<channel-type id="batteryavailableenergy">
 		<item-type>Number:Energy</item-type>
 		<label>Battery Energy Available</label>
-		<state pattern="%.1f %unit%" readOnly="true"></state>
+		<state pattern="%.1f kWh" readOnly="true"></state>
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml:100]: kWh is default unit for energy, I believe - is this change needed?

File: bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/api/Car.java [hunk 8]
```diff
@@ -105,6 +114,9 @@ public void setBatteryStatus(JsonObject responseJson) {
                 if (attributes.get("chargingRemainingTime") != null) {
                     chargingRemainingTime = attributes.get("chargingRemainingTime").getAsInt();
                 }
+                if (attributes.get("timestamp") != null) {
+                    batteryStatusUpdated = attributes.get("timestamp").getAsString();
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/api/Car.java:118]: Which format does this timestamp string have? Perhaps it would be better to convert it to an `Instant` or `ZonedDateTime` here, so callers of `getBatteryStatusUpdated` won't have to know the format and deal with it.

File: bundles/org.openhab.binding.renault/README.md [hunk 9]
```diff
@@ -69,18 +72,21 @@ renaultcar.sitemap:
...
-        Default item=RenaultCar_EstimatedRange
-        Default item=RenaultCar_Odometer
-        Selection item=RenaultCar_HVACStatus mappings=[ON="ON"] icon="switch"
-        Setpoint item=RenaultCar_HVACTargetTemperature minValue=19 maxValue=21 step=1 icon="temperature"
-        Default item=RenaultCar_LocationUpdate icon="time"
-        Default item=RenaultCar_Location
+        Switch icon="switch" item=RenaultCar_UpdateNow
+        Default icon="batterylevel" item=RenaultCar_BatteryLevel
+        Default icon="energy" item=RenaultCar_BatteryEnergyAvailable
+        Default icon="time" item=RenaultCar_BatteryStatusUpdated
```
Comment (jlaur) [inline bundles/org.openhab.binding.renault/README.md:78]: icon wouldn't be needed here if it was added to the channel type, since created items would then automatically inherit it.

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for the improvements. I have added some quick feedback. Please also consider if the sitemap example is strictly needed. It's clearly obsoleted by now, but it seems each update will add ~180 kB to the repository.

File: bundles/org.openhab.binding.renault/README.md (hunk 1 above)
Comment (dougculnane) [reply bundles/org.openhab.binding.renault/README.md:30]: I am not sure how best to document this.  I will add a Default column with values to see if this is clearer?

File: bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml (hunk 7 above)
Comment (dougculnane) [reply bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml:100]: Not sure but I got the converted value in J (for Joules) in my site map page so I added the unit and it displayed the values and unity in kWh which is what I want.   Maybe there is a better way to define the units?

Comment (jlaur) [reply bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml:100]: Do you also get this with final release version 3.4? There was a fix in core merged into one of the last milestones: openhab/openhab-core#3201.

File: bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/handler/RenaultHandler.java (hunk 4 above)
Comment (dougculnane) [reply bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/handler/RenaultHandler.java:203]: @jlaur Thanks for the feedback.  Sorry it is hard for me to find time at the moment to process it.  I did however investigate the RefreshType.  It seams to me that this is used by the UI to refresh the state of items etc... Is it possible to trigger a binding refresh from a rule?  The problem we wish to solve is described here: https://community.openhab.org/t/betatest-renault-ze-services-binding/72226/222?u=doug_culnane

"Now that we have 2 cars and only one wallbox, I’d like to know which of the car is plugged to the charger.
With the automated polling I will receive this information after at least 5 minutes (in my case), that could be to long.

I have a workaround for this, but its also a little complicated.
I disable the car Thing with an HttpPutRequest - wait for 1 second - and enable the car thing again with the HttpPutRequest."

The "updatenow" channel is a better work around that this.  However maybe there is a Item.refresh() function that I could implement on the channel and this would work better with the openHAB platform?

File: bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml (hunk 7 above)
Comment (dougculnane) [reply bundles/org.openhab.binding.renault/src/main/resources/OH-INF/thing/thing-types.xml:100]: I will test this. I did get this issue openhab:3.4.0.M5.

File: bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/handler/RenaultHandler.java (hunk 4 above)
Comment (jlaur) [reply bundles/org.openhab.binding.renault/src/main/java/org/openhab/binding/renault/internal/handler/RenaultHandler.java:203]: > Is it possible to trigger a binding refresh from a rule?

Yes, for example in a DSL rule:
```java
sendCommand(itemName, "REFRESH")
```

**Comment by jlaur** [comment]
@dougculnane - I'm afraid you just performed some git operation so that 656 files are changed across many bindings, and review is requested from many contributors. Perhaps this could be helpful: https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

**Comment by dougculnane** [comment]
Sorry.  I used the web UI to commit suggestion commits from the review.  These commits we not sighed off by me so when I merged and updated the pull request the checks failed.  I followed the rebase instructions and it added my signature to all commits. This is not a correct clearly. 

Sorry for any inconvenience.  I will close this pull request and resubmit correctly later with the feedback from @jlaur

**Comment by jlaur** [comment]
@dougculnane - if you can reuse the branch name and force push it when you have fixed the issue, this PR could be reopened to preserve history.

**Comment by dougculnane** [comment]
> @dougculnane - if you can reuse the branch name and force push it when you have fixed the issue, this PR could be reopened to preserve history.

@jlaur I am not 100% sure how to rebase this without experimenting and probably generating lots of spam to the list of Reviewers.  I can not remove them and updating my branch (to remove the signed commits) did not update this pull request.  So will leave this closed and start a new pull request. Sorry for the mess.
